File: src/main/java/ffblockly/meta/FFBlockBuilder.java
package ffblockly.meta;

import java.text.MessageFormat;
import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import static java.util.Arrays.asList;
import static java.util.stream.Stream.concat;
import static ffblockly.meta.FFBlocklyInputValueBuilder.FFBFieldType.input_statement;

public interface FFBlockBuilder {

    int ORDINAL = input_statement.ordinal();

    static FFBlock createFFBlock(final FilterListing filterListing) {
        return Scan.MYFACTORY.block(new FFBlock() {
            boolean loaded;
            private List<AvOption> options;

            @Override
            public Boolean isInputsInline() {
                return true;
            }

            @Override
            public FilterListing.FilterSignature getPreviousStatement() {
                return filterListing.getInputSignature().iterator().hasNext() ? filterListing.getInputSignature().get(0) : null;
            }

            @Override
            public FilterListing.FilterSignature getNextStatement() {
                return filterListing.getReturnSignature().iterator().hasNext() ? filterListing.getReturnSignature().get(0) : null;
            }

            @Override
            public int getColour() {
                final List<FilterListing.FilterSignature> inputSignature = filterListing.getInputSignature() ;
                return  inputSignature.isEmpty()?0:FFBlockly.colorForString(String.valueOf(inputSignature.iterator().next()));
            }

            @Override
            public String getTooltip() {
                return filterListing.getShortDescription();
            }

            @Override
            public String getHelpUrl() {
                return MessageFormat.format("https://ffmpeg.org/ffmpeg-filters.html#{0}", filterListing.getFilterName());
            }

            @Override
            public String getType() {
                return filterListing.getFilterName();
            }

            @Override
            public String getMessage0() {
                final StringBuilder msg0 = new StringBuilder(this.getType());
                final List<FFBlocklyFieldValue> args0 = this.getArgs0();
                boolean first = true;

                if (null != args0) for (int i = 0; i < args0.size(); i++) {
                     FFBlocklyFieldValue theArg = args0.get(i);

                    if (null != theArg.getType()) {
                        FFBlocklyInputValueBuilder.FFBFieldType type = theArg.getType();
                        if (type.ordinal() < input_statement.ordinal()) {
                            msg0.append(first ? "=" : ":").append(theArg.getName()).append("=%").append(i + 1);
                            first = false;
                        }
                    }
                }
                return msg0.toString();
            }

            @Override
            public List<FFBlocklyFieldValue> getArgs0() {

                int[] c = {1};

                final Function<FilterListing.FilterSignature, FFBlocklyFieldValue> filterSignatureFFBlocklyFieldValueFunction = filterSignature -> FFBlockBuilder.getFfBlocklyFieldValue(c, filterSignature);

                final Stream<FFBlocklyFieldValue> stream1 = filterListing.getInputSignature().stream().skip(1).map(filterSignatureFFBlocklyFieldValueFunction);


                if (!this.loaded) {
                    this.loaded = true;
                    this.options = filterListing.getOptions();
                }
                c[0] = 1;
                final List<FFBlocklyFieldValue> collect = null;
                Stream<FFBlocklyFieldValue> ffBlocklyFieldValueStream = null == this.options ? new ArrayList<FFBlocklyFieldValue>().stream() : this.options.stream().map(FFBlocklyInputValueBuilder::createFFBlocklyInputValue);

                final List<FFBlocklyFieldValue> ret = concat(stream1, concat(concat(ffBlocklyFieldValueStream, asList(Scan.MYFACTORY.fil(new FFBlocklyFieldValue() {
                    @Override
                    public String getValue() {
                        return null;
                    }

                    @Override
                    public FFBlocklyInputValueBuilder.FFBFieldType getType() {
                        return FFBlocklyInputValueBuilder.FFBFieldType.input_dummy;
                    }

                    @Override
                    public String getText() {
                        return null;
                    }

                    @Override
                    public String getMin() {
                        return null;
                    }

                    @Override
                    public String getMax() {
                        return null;
                    }

                    @Override
                    public Boolean getChecked() {
                        return null;
                    }

                    @Override
                    public List<List<String>> getOptions() {
                        return null;
                    }

                    @Override
                    public String getName() {
                        return null;
                    }

                    @Override
                    public String getTooltip() {
                        return null;
                    }

                    @Override
                    public String getCheck() {
                        return null;
                    }
                }).as()).stream()), filterListing.getReturnSignature().stream().skip(1).map(filterSignatureFFBlocklyFieldValueFunction))).collect(Collectors.toList());


                return ret.isEmpty() ? null : ret;
            }
        }).as();
    }

    static FFBlocklyFieldValue getFfBlocklyFieldValue(int[] c, FilterListing.FilterSignature filterSignature) {
        c[0]++;
        return Scan.MYFACTORY.fil(new FFBlocklyFieldValue() {
            @Override
            public String getValue() {
                return null;
            }

            @Override
            public FFBlocklyInputValueBuilder.FFBFieldType getType() {
                return input_statement;
            }

            @Override
            public String getText() {
                return null;
            }

            @Override
            public String getMin() {
                return null;
            }

            @Override
            public String getMax() {
                return null;
            }

            @Override
            public Boolean getChecked() {
                return null;
            }

            @Override
            public List<List<String>> getOptions() {
                return null;
            }

            @Override
            public String getName() {
                return "Input" + filterSignature.name() + c[0]++;
            }

            @Override
            public String getTooltip() {
                return null;
            }

            @Override
            public String getCheck() {
                return null;
            }
        }).as();
    }
}

File: src/main/java/ffblockly/meta/FFBlockBuilder.kt
package ffblockly.meta

import ffblockly.meta.FFBlockly.Companion.colorForString
import ffblockly.meta.FFBlocklyInputValueBuilder.*
import ffblockly.meta.FilterListing.FilterSignature
import java.text.MessageFormat
import java.util.function.Function

interface FFBlockBuilder {
    companion object {   val inputDummy = object : FFBlocklyFieldValue {
        override val value: String?
            get() = null

        override val type: FFBFieldType
            get() = FFBFieldType.input_dummy

        override val text: String?
            get() = null

        override val min: String?
            get() = null

        override val max: String?
            get() = null

        override val checked: Boolean?
            get() = null

        override val options: List<List<String>>?
            get() = null

        override val name: String?
            get() = null

        override val tooltip: String?
            get() = null

        override val check: String?
            get() = null
    }
        @JvmStatic
        fun createFFBlock(filterListing: FilterListing): FFBlock? {
            return Scan.MYFACTORY.block(object : FFBlock {
                var loaded = false
                private var options: List<AvOption>? = null
                override val isInputsInline: Boolean
                    get() = true

                override val previousStatement: FilterSignature?
                    get() = if (filterListing.inputSignature!!.iterator().hasNext()) filterListing.inputSignature!![0] else null

                override val nextStatement: FilterSignature?
                    get() = if (filterListing.returnSignature!!.iterator().hasNext()) filterListing.returnSignature!![0] else null

                override val colour: Int
                    get() {
                        val inputSignature = filterListing.inputSignature
                        return if (inputSignature!!.isEmpty()) 0 else colorForString(inputSignature.iterator().next().toString())
                    }

                override val tooltip: String?
                    get() = filterListing.shortDescription

                override val helpUrl: String
                    get() =  "https://ffmpeg.org/ffmpeg-filters.html#${filterListing.filterName}"

                override val type: String
                    get() = filterListing.filterName

                override val message0: String
                    get() {
                        var msg0 = StringBuilder(type)
                        val args0 = args0
                        var first = true
                        if (null != args0) args0.indices.forEach { i ->
                            val theArg = args0[i]
                            if (null != theArg.type) {
                                val type = theArg.type
                                when {
                                    type!!.ordinal < FFBFieldType.input_statement.ordinal -> {
                                        msg0.append(if (first) "=" else ":").append(theArg.name).append("=%").append(i + 1)
                                        first = false
                                    }
                                     type!!.ordinal ==FFBFieldType.input_statement.ordinal->{
                                      msg0=java.lang.StringBuilder   ("[${theArg.name}]%${ i + 1}$msg0")
                                     }
                                }
                            }
                        }
                        return msg0.toString()
                    }

                private val autoBean = Scan.MYFACTORY.fil(inputDummy)

                override val args0: List<FFBlocklyFieldValue>?
                    get() {
                        val c = intArrayOf(1)
                        val filterSignatureFFBlocklyFieldValueFunction = Function { filterSignature: FilterSignature -> getFfBlocklyFieldValue( c, filterSignature) }
                        val inputSignature = filterListing.inputSignature
                        val list3 = inputSignature!! .drop(1)?.map  { filterSignature ->  filterSignatureFFBlocklyFieldValueFunction.apply(filterSignature  ) }
                        if (!loaded) {
                            loaded = true
                            options = filterListing.options
                        }
                         c[0] = 1

                      return   options?.  map(FFBlocklyInputValueBuilder.Companion::createFFBlocklyInputValue)?.let {

                          val list = filterListing.returnSignature!!
                          val list1 = list.drop(1).map( filterSignatureFFBlocklyFieldValueFunction::apply)
                          val list2 =  (it + list1)
                          (list3 + list2).filterNotNull()
                      }
                    }
            }).`as`()
        }

        fun getFfBlocklyFieldValue(c: IntArray, filterSignature: FilterSignature): FFBlocklyFieldValue? {
            c[0]++
            return Scan.MYFACTORY.fil(object : FFBlocklyFieldValue {
                override val value: String?
                    get() = null

                override val type: FFBFieldType
                    get() = FFBFieldType.input_statement

                override val text: String?
                    get() = null

                override val min: String?
                    get() = null

                override val max: String?
                    get() = null

                override val checked: Boolean?
                    get() = null

                override val options: List<List<String>>?
                    get() = null

                override val name: String
                    get() = "Input" + filterSignature.name + c[0]++

                override val tooltip: String?
                    get() = null

                override val check: String?
                    get() = null
            }).`as`()
        }

        val ORDINAL = FFBFieldType.input_statement.ordinal
    }
}